cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_svd_impl_dpc.cpp

```diff
@@ -0,0 +1,207 @@
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+namespace oneapi::dal::pca::backend {
+
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+using alloc = sycl::usm::alloc;
+
+using bk::context_gpu;
+using model_t = model<task::dim_reduction>;
+using task_t = task::dim_reduction;
+using input_t = train_input<task_t>;
+using result_t = train_result<task_t>;
+using descriptor_t = detail::descriptor_base<task_t>;
+
+template <typename Float>
+auto compute_eigenvectors_on_host(sycl::queue& q,
+                                  pr::ndarray<Float, 2>&& data,
+                                  std::int64_t component_count,
+                                  const dal::backend::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(compute_eigenvectors_on_host);
+    ONEDAL_ASSERT(data.has_mutable_data());
+    ONEDAL_ASSERT(data.get_dimension(0) == data.get_dimension(1),
+                  "dataelation matrix must be square");
+    ONEDAL_ASSERT(data.get_dimension(0) > 0);
+    const std::int64_t column_count = data.get_dimension(0);
+
+    auto eigvecs = pr::ndarray<Float, 2>::empty({ component_count, column_count });
+    auto eigvals = pr::ndarray<Float, 1>::empty(component_count);
+
+    auto host_data = data.to_host(q, deps);
+    pr::sym_eigvals_descending(host_data, component_count, eigvecs, eigvals);
+
+    return std::make_tuple(eigvecs, eigvals);
+}
+
+template <typename Float>
+auto centered_data(sycl::queue& q,
+                   const pr::ndview<Float, 2>& data,
+                   const bk::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(centered_data, q);
+    ONEDAL_ASSERT(data.has_data());
+    ONEDAL_ASSERT(data.get_dimension(1) > 0);
+    const std::int64_t column_count = data.get_dimension(1);
+    const std::int64_t row_count = data.get_row_count();
+
+    pr::ndarray<Float, 2> data_copy =
+        pr::ndarray<Float, 2>::copy(data, { row_count, column_count });
+
+    auto event = q.submit([&](sycl::handler& cgh) {
+        auto data_acc = data_copy.get_mutable_data();
+
+        cgh.parallel_for(sycl::range<1>(column_count), [=](sycl::id<1> idx) {
+            Float sum = 0.0;
+            for (std::int64_t i = 0; i < row_count; i++) {
+                sum += data_acc[i * column_count + idx[0]];
+            }
+            Float mean = sum / row_count;
+            for (std::int64_t i = 0; i < row_count; i++) {
+                data_acc[i * column_count + idx[0]] -= mean;
+            }
+        });
+    });
+
+    return std::make_tuple(data_acc, event);
+}
+
+template <typename Float>
+auto svd_decomposition(sycl::queue& q,
+                       pr::ndview<Float, 2>& data,
+                       std::size_t row_count,
+                       std::size_t column_count) {
+    auto U = pr::ndarray<Float, 2>::empty({ row_count, row_count });
+    auto S = pr::ndarray<Float, 1>::empty({ column_count });
+    auto V_T = pr::ndarray<Float, 2>::empty({ column_count, column_count });
```

Comment:
1) Here you have implicit conversions
2) You probably want to allocate this arrays not on host, plus using USM memory